src/wallet/scriptpubkeyman.cpp

```diff
@@ -387,31 +456,15 @@ bool LegacyScriptPubKeyMan::SetHDChain(WalletBatch &batch, const CHDChain& chain
         return false;
 
     if (!memonly) {
-        if (!batch.WriteHDChain(chain)) {
-            throw std::runtime_error(std::string(__func__) + ": WriteHDChain failed");
-        }
-
-        m_storage.UnsetBlankWalletFlag(batch);
-    }
-
-    return true;
-}
-
-bool LegacyScriptPubKeyMan::SetCryptedHDChain(WalletBatch &batch, const CHDChain& chain, bool memonly)
-{
-    LOCK(cs_KeyStore);
-
-    if (!SetCryptedHDChain(chain))
-        return false;
-
-    if (!memonly) {
-        if (encrypted_batch) {
-            if (!encrypted_batch->WriteCryptedHDChain(chain))
-                throw std::runtime_error(std::string(__func__) + ": WriteCryptedHDChain failed");
+        if (chain.IsCrypted() && encrypted_batch) {
+            if (!encrypted_batch->WriteHDChain(chain))
+                throw std::runtime_error(std::string(__func__) + ": WriteHDChain failed");
         } else {
-            if (!batch.WriteCryptedHDChain(chain))
-                throw std::runtime_error(std::string(__func__) + ": WriteCryptedHDChain failed");
+            if (!batch.WriteHDChain(chain)) {
+                throw std::runtime_error(std::string(__func__) + ": WriteHDChain failed");
```

Comment:
nit: adjust error messages to make it easier to see which one (encrypted or not) failed exactly